### PR TITLE
Add dependent schema relationship type

### PR DESCRIPTION
**Pull request by laurenleach**
Add dependent schema relationship type

### CODE REVIEW

**Review by ggreer — APPROVED** [review]

File: pkg/field/relationships.go [hunk 1]
```diff
@@ -8,11 +8,65 @@ const (
...
 )
 
+type SchemaFieldRelationshipI interface {
+	ValidateConstraint(fieldsPresent map[string]int) error
+}
+
+type DependentSchemaFieldRelationship struct {
+	RequiredFields  []SchemaField
+	DependentFields []SchemaField
+}
```
Comment (shackra) [inline pkg/field/relationships.go:18]: @laurenleach I don't follow why a new struct for tracking required fields and dependent fields (of those required fields?), why add a new `const` below `AtLeastOne` and extend `SchemaFieldRelationship` with a new field?

The problem with array/slice of an interface is that `[]DependentSchemaFieldRelationship` or `[]SchemaFieldRelationship` doesn't implement a `[]SchemaFieldRelationshipI`

Comment (ggreer) [reply pkg/field/relationships.go:18]: I like this idea.

Comment (shackra) [reply pkg/field/relationships.go:18]: my objection comes from this btw https://stackoverflow.com/questions/44319906/why-golang-struct-array-cannot-be-assigned-to-an-interface-array

Comment (laurenleach) [reply pkg/field/relationships.go:18]: why do we need `[]DependentSchemaFieldRelationship` or `[]SchemaFieldRelationship`? we can use `[]SchemaFieldRelationshipI`

Comment (shackra) [reply pkg/field/relationships.go:18]: `SchemaFieldRelationshipI` is needed in so far as changes to the previous API were introduced, my objection is twofold:

1. We don't know how users outside C1 will try to use the SDK
2. We can still extend the concrete type `SchemaField` to achieve the new relationship constraint

Regarding nº 1 some users can find [this pitfall](https://stackoverflow.com/questions/44319906/why-golang-struct-array-cannot-be-assigned-to-an-interface-array) when developing their own connectors relaying on the introduced interface `SchemaFieldRelationshipI`.

With nº 2 I've opened a PR showing that extending what is already there can help https://github.com/ConductorOne/baton-sdk/pull/187

**Pull request merged**